lib/rubygems/remote_fetcher.rb

```diff
@@ -25,13 +26,10 @@ class FetchError < Gem::Exception
     attr_accessor :uri, :original_uri
 
     def initialize(message, uri)
-      super message
-
-      uri = Gem::UriParser.parse_uri(uri)
-
       @original_uri = uri.dup
+      uri = Gem::PrintableUri.parse_uri(uri)
 
-      uri.password = 'REDACTED' if uri.respond_to?(:password) && uri.password
+      super(uri.valid_uri? && uri.original_password ? message.sub(uri.original_password, 'REDACTED') : message)
```

Comment:
Sorry I keep coming up with improvements. I think it'd be neat to move this to `Gem::PrintableUri`, to keep all redacting functionality together? Maybe something like

```ruby
def redact_credentials_from(text)
  return text unless valid_uri? && original_password?

  text.sub(uri.original_password, 'REDACTED')
end
```

Reply:
By doing the above, we no longer need to expose an `original_password` method.

Reply:
I did this as part of https://github.com/rubygems/rubygems/pull/4868/commits/9581c2740a37ef12ac0351cb9f166e3d6aa448c1.